lib/utils/analytics/index.js

```diff
@@ -29,147 +28,160 @@ const logError = (type, error) => {
+  }
+
+  if (!id) return processResponseBody(response, id, startTime);
+
+  const cachePath = join(cacheDirPath, id);
+  try {
+    await fse.unlink(cachePath);
+  } catch (error) {
+    logError(`Could not remove cache file ${id}`, error);
+  }
+
+  return processResponseBody(response, id, startTime);
+}
+
+async function report(payload, options = {}) {
+  ensurePlainObject(payload);
+  if (!analyticsUrl) return null;
+  const isForced = options && options.isForced;
+  if (areAnalyticsDisabled && !isForced) return null;
+  if (!cacheDirPath) return request(payload);
+  const id = uuid();
+
+  const [, requestResult] = await Promise.all([
+    (async function self() {
+      try {
+        return await fse.writeJson(join(cacheDirPath, id), { payload, timestamp: Date.now() });
+      } catch (error) {
+        if (error.code === 'ENOENT') {
+          try {
+            await fse.ensureDir(cacheDirPath);
+            return self();
+          } catch (ensureDirError) {
+            logError('Cache dir creation error:', ensureDirError);
           }
-        );
-      },
-      (networkError) => {
-        logError('Request network error', networkError);
+        }
+        logError(`Write cache file error: ${id}`, error);
         return null;
       }
-    );
-  });
+    })(),
+    request(payload, { id }),
+  ]);
+  return requestResult;
 }
 
-function report(payload, options = {}) {
-  return BbPromise.try(() => {
-    ensurePlainObject(payload);
-    if (!analyticsUrl) return null;
-    const isForced = options && options.isForced;
-    if (areAnalyticsDisabled && !isForced) return null;
-    if (!cacheDirPath) return request(payload);
-    const id = uuid();
-    return BbPromise.all([
-      (function self() {
-        return fse
-          .writeJson(join(cacheDirPath, id), { payload, timestamp: Date.now() })
-          .catch((error) => {
-            if (error.code === 'ENOENT') {
-              return fse.ensureDir(cacheDirPath).then(self, (ensureDirError) => {
-                logError('Cache dir creation error:', ensureDirError);
-              });
-            }
-            logError(`Write cache file error: ${id}`, error);
-            return null;
-          });
-      })(),
-      request(payload, { id }),
-    ]).then(([, requestResult]) => requestResult); // In all cases resolve with request result
-  });
-}
+async function sendPending(options = {}) {
+  const isForced = options && options.isForced;
+  serverlessRunEndTime = null; // Needed for testing
+  if (options.serverlessExecutionSpan) {
+    options.serverlessExecutionSpan.then(markServerlessRunEnd, markServerlessRunEnd);
+  }
+  if (areAnalyticsDisabled && !isForced) return null;
+  if (!cacheDirPath) return null;
+  const limit = pLimit(3);
+  let dirFilenames;
+  try {
+    dirFilenames = await fse.readdir(cacheDirPath);
+  } catch (readdirError) {
+    if (readdirError.code !== 'ENOENT') logError('Cannot access cache dir', readdirError);
+    return null;
```

Comment:
We may just `return` (no need to `return null`, same below. In old version it was `null`, to visually better signal intent (`then(() => {}` or `then(() => { return undefined }` is either confusing or dirty)